[![Build Status](https://travis-ci.org/gVkWY8NJAa/ProFootballRef.svg?branch=master)](https://travis-ci.org/gVkWY8NJAa/ProFootballRef) [![Coverage Status](https://coveralls.io/repos/github/gVkWY8NJAa/ProFootballRef/badge.svg?branch=master)](https://coveralls.io/github/gVkWY8NJAa/ProFootballRef?branch=master)
# ProFootballRef </hr>

This is a python toolkit that lets you scrape statistics from https://www.pro-football-reference.com/, and return the resulting data as a Pandas DataFrame. 

Please consider contributing the $20/yr to support the site, they do a great job: https://www.pro-football-reference.com/my/?do=ad_free_browsing

## Contents
* [Installation](#installation)
* [Testing](#testing)
* [Find players](#find_players)
* [Player stats](#player_stats)
    * [Individual player stats per season](#career_player_stats)
    * [Multiple player stats for a given season](#multi_player_stats)
    * [Multiple player stats for multiple seasons](#multi_player_multi_season)
    * [Gamelog](#gamelog)
* [Team stats](#team_stats)
    * [Team offense stats](#team_offense)
    * [Team defense stats](#team_defense)

## Key Features
* Aggregate player data for each season.
* Ability to combine qualitative (height/weight) with quantitative (TDs).
* Multi column headers have been simplified and closly match the canonical source.
* Scrape team stats for a given season.
* Player gamelog data available for a given season.
* Returned objects are Pandas DataFrames for ease of analysis.

<a id='installation'></a>
## Installation
```
git clone git@github.com:gVkWY8NJAa/ProFootballRef.git
cd ProFootballRef
pip install -r requirements.txt
```
<a id='testing'></a>
## Testing
```
cd <path/to/ProFootballRef>
python3.6 -m pytest tests/
```
<a id='find_players'></a>
## Find Players
---
Before we do anything, we need to gather a list of urls for various players to parse.

This is easily done by importing the **GetPositionLinks** module that resides in the 'LinkBuilder' directory:

In [2]:
from profootballref.LinkBuilder import GetPositionLinks

The **GetPositionLinks** module contains a class called **Position**. 

To generate urls to parse, we'll call the **player_links** method from the **Position** class, and save the output to a list.
The **Position** class takes one of five possible arguments:
* passing
* receiving
* rushing
* kicking
* defense

The **player_links** method takes either a season, or a range of seasons, as integers:

In [3]:
urls = GetPositionLinks.Position('passing').player_links(2017)

or to grab all players from a range of seaons do:

In [4]:
urls = GetPositionLinks.Position('passing').player_links(2015, 2017)

If a range of seasons is passed to **player_links**, any duplicate urls will be removed automatically.

We'll look at the first five urls in the list:

In [3]:
print(urls[:5])

['https://www.pro-football-reference.com/players/B/BradTo00.htm', 'https://www.pro-football-reference.com/players/R/RivePh00.htm', 'https://www.pro-football-reference.com/players/M/MannEl00.htm', 'https://www.pro-football-reference.com/players/S/StafMa00.htm', 'https://www.pro-football-reference.com/players/R/RoetBe00.htm']


<a id='player_stats'></a>
## Player stats 
---
The following code demonstrates how to return career position statistics given a player. This is the data that would be found on the [players page](https://www.pro-football-reference.com/players/B/BradTo00.htm).
<a id='career_player_stats'></a>
### Individual player stats
In this example, we will [pass a url as a string for a given player](#find_players) that we created previously to return their career stats for their position. If you do not yet have a list of urls for a given position, see the above section [Find Players](#find_players).

This is easily done by importing the **PlayerParser** module that resides in the 'Parsers' directory:

In [4]:
from profootballref.Parsers import PlayerParser

The **PlayerParser** module contains a class also called **PlayerParser**. 

To scrape a player(s) career stats, we'll call one of five methods from the **PlayerParser** class, and save the output to a variable.

The available methods from the **PlayerParser** class are:
* passing
* receiving
* rushing
* kicking
* defense

Remember we're using the 'urls' data from what we did in the [Find Players](#find_players) section above.

In [5]:
passing_df = PlayerParser.PlayerParser().passing(urls[:1][0])

In [6]:
passing_df.head()

Name  Year   Age Throws  Height  Weight  DOB_mo  DOB_day  DOB_yr  \
0  Tom Brady  2001  24.0  Right      76     225       8        3    1977   
1  Tom Brady  2002  25.0  Right      76     225       8        3    1977   
2  Tom Brady  2003  26.0  Right      76     225       8        3    1977   
3  Tom Brady  2004  27.0  Right      76     225       8        3    1977   
4  Tom Brady  2005  28.0  Right      76     225       8        3    1977   

    College  ...  Rec_Yds   Y/R  Rec_TD  Rec_Lng  R/G Rec_Y/G   Ctch%  YScm  \
0  Michigan  ...     23.0  23.0     0.0     23.0  0.1     1.5  100.0%    66   
1  Michigan  ...      NaN   NaN     NaN      NaN  NaN     NaN     NaN   110   
2  Michigan  ...      NaN   NaN     NaN      NaN  NaN     NaN     NaN    63   
3  Michigan  ...      NaN   NaN     NaN      NaN  NaN     NaN     NaN    28   
4  Michigan  ...      NaN   NaN     NaN      NaN  NaN     NaN     NaN    89   

   RRTD   Fmb  
0     0  12.0  
1     1  11.0  
2     1  13.0  
3     0   7.0  
4     1   4.0  

[5 rows x 58 columns]

<a id='multi_player_stats'></a>
### Multiple player stats
**This can generate a ton of traffic to the website so use caution with how many players you retrieve at one time.**

In [1]:
import pandas as pd
from profootballref.LinkBuilder import GetPositionLinks
from profootballref.Parsers import PlayerParser
 
# Initialize an empty DataFrame to store all the players
all_qb = pd.DataFrame()
 
# Specify which position and season we want
position = 'passing'

In [16]:
# In this example we'll generate a list of urls multiple seasons
links = GetPositionLinks.Position(position).player_links(2013,2015)

In [2]:
import random

In [3]:
links = ['https://www.pro-football-reference.com/players/M/MannPe00.htm',
 'https://www.pro-football-reference.com/players/R/RyanMa00.htm',
 'https://www.pro-football-reference.com/players/B/BreeDr00.htm',
 'https://www.pro-football-reference.com/players/S/StafMa00.htm',
 'https://www.pro-football-reference.com/players/B/BradTo00.htm',
 'https://www.pro-football-reference.com/players/F/FlacJo00.htm',
 'https://www.pro-football-reference.com/players/T/TannRy00.htm',
 'https://www.pro-football-reference.com/players/D/DaltAn00.htm',
 'https://www.pro-football-reference.com/players/R/RoetBe00.htm',
 'https://www.pro-football-reference.com/players/P/PalmCa00.htm',
 'https://www.pro-football-reference.com/players/L/LuckAn00.htm',
 'https://www.pro-football-reference.com/players/M/MannEl00.htm',
 'https://www.pro-football-reference.com/players/R/RivePh00.htm',
 'https://www.pro-football-reference.com/players/R/RomoTo00.htm',
 'https://www.pro-football-reference.com/players/S/SmitAl03.htm',
 'https://www.pro-football-reference.com/players/H/HennCh01.htm',
 'https://www.pro-football-reference.com/players/N/NewtCa00.htm',
 'https://www.pro-football-reference.com/players/G/GrifRo01.htm',
 'https://www.pro-football-reference.com/players/S/SmitGe00.htm',
 'https://www.pro-football-reference.com/players/G/GlenMi00.htm',
 'https://www.pro-football-reference.com/players/K/KaepCo00.htm',
 'https://www.pro-football-reference.com/players/W/WilsRu00.htm',
 'https://www.pro-football-reference.com/players/S/SchaMa00.htm',
 'https://www.pro-football-reference.com/players/C/CutlJa00.htm',
 'https://www.pro-football-reference.com/players/F/FitzRy00.htm',
 'https://www.pro-football-reference.com/players/C/CampJa00.htm',
 'https://www.pro-football-reference.com/players/F/FoleNi00.htm',
 'https://www.pro-football-reference.com/players/M/ManuEJ00.htm',
 'https://www.pro-football-reference.com/players/R/RodgAa00.htm',
 'https://www.pro-football-reference.com/players/P/PryoTe00.htm',
 'https://www.pro-football-reference.com/players/W/WeedBr00.htm',
 'https://www.pro-football-reference.com/players/B/BradSa00.htm',
 'https://www.pro-football-reference.com/players/C/CassMa00.htm',
 'https://www.pro-football-reference.com/players/K/KeenCa00.htm',
 'https://www.pro-football-reference.com/players/C/ClemKe00.htm',
 'https://www.pro-football-reference.com/players/P/PondCh00.htm',
 'https://www.pro-football-reference.com/players/M/McCoJo01.htm',
 'https://www.pro-football-reference.com/players/M/McGlMa01.htm',
 'https://www.pro-football-reference.com/players/F/FlynMa00.htm',
 'https://www.pro-football-reference.com/players/L/LockJa00.htm',
 'https://www.pro-football-reference.com/players/L/LewiTh01.htm',
 'https://www.pro-football-reference.com/players/C/CousKi00.htm',
 'https://www.pro-football-reference.com/players/F/FreeJo00.htm',
 'https://www.pro-football-reference.com/players/V/VickMi00.htm',
 'https://www.pro-football-reference.com/players/H/HoyeBr00.htm',
 'https://www.pro-football-reference.com/players/T/TolzSc00.htm',
 'https://www.pro-football-reference.com/players/G/GabbBl00.htm',
 'https://www.pro-football-reference.com/players/T/TuelJe00.htm',
 'https://www.pro-football-reference.com/players/O/OrtoKy00.htm',
 'https://www.pro-football-reference.com/players/B/BarkMa00.htm',
 'https://www.pro-football-reference.com/players/D/DaniCh00.htm',
 'https://www.pro-football-reference.com/players/S/SimmMa00.htm',
 'https://www.pro-football-reference.com/players/W/WallSe00.htm',
 'https://www.pro-football-reference.com/players/Y/YateT.00.htm',
 'https://www.pro-football-reference.com/players/O/OsweBr00.htm',
 'https://www.pro-football-reference.com/players/P/PainCu00.htm',
 'https://www.pro-football-reference.com/players/J/JackTa00.htm',
 'https://www.pro-football-reference.com/players/H/HassMa00.htm',
 'https://www.pro-football-reference.com/players/D/DaviDo02.htm',
 'https://www.pro-football-reference.com/players/M/MoorMa01.htm',
 'https://www.pro-football-reference.com/players/T/TaylTy00.htm',
 'https://www.pro-football-reference.com/players/C/CribJo01.htm',
 'https://www.pro-football-reference.com/players/K/koenemic01.htm',
 'https://www.pro-football-reference.com/players/A/AustTa00.htm',
 'https://www.pro-football-reference.com/players/B/BrowAn04.htm',
 'https://www.pro-football-reference.com/players/B/BushJo00.htm',
 'https://www.pro-football-reference.com/players/F/FitzLa00.htm',
 'https://www.pro-football-reference.com/players/H/HekkJo00.htm',
 'https://www.pro-football-reference.com/players/J/JameMi00.htm',
 'https://www.pro-football-reference.com/players/D/DrewMa00.htm',
 'https://www.pro-football-reference.com/players/K/KerlJe00.htm',
 'https://www.pro-football-reference.com/players/L/LannSp00.htm',
 'https://www.pro-football-reference.com/players/M/McBrMa20.htm',
 'https://www.pro-football-reference.com/players/M/McCoLu00.htm',
 'https://www.pro-football-reference.com/players/M/McCoCo00.htm',
 'https://www.pro-football-reference.com/players/M/McFaDa00.htm',
 'https://www.pro-football-reference.com/players/P/PetePa00.htm',
 'https://www.pro-football-reference.com/players/P/PoweBi00.htm',
 'https://www.pro-football-reference.com/players/R/RainBo00.htm',
 'https://www.pro-football-reference.com/players/R/ReecMa00.htm',
 'https://www.pro-football-reference.com/players/R/RobiDe00.htm',
 'https://www.pro-football-reference.com/players/S/SandAc00.htm',
 'https://www.pro-football-reference.com/players/S/SanuMo00.htm',
 'https://www.pro-football-reference.com/players/S/SchiSh99.htm',
 'https://www.pro-football-reference.com/players/S/SmitBr01.htm',
 'https://www.pro-football-reference.com/players/C/CarrDe02.htm',
 'https://www.pro-football-reference.com/players/B/BortBl00.htm',
 'https://www.pro-football-reference.com/players/B/BridTe00.htm',
 'https://www.pro-football-reference.com/players/S/SancMa00.htm',
 'https://www.pro-football-reference.com/players/D/DaviAu00.htm',
 'https://www.pro-football-reference.com/players/S/StanDr00.htm',
 'https://www.pro-football-reference.com/players/H/HillSh00.htm',
 'https://www.pro-football-reference.com/players/W/WhitCh02.htm',
 'https://www.pro-football-reference.com/players/M/MettZa00.htm',
 'https://www.pro-football-reference.com/players/A/AndeDe00.htm',
 'https://www.pro-football-reference.com/players/L/LindRy00.htm',
 'https://www.pro-football-reference.com/players/M/MallRy00.htm',
 'https://www.pro-football-reference.com/players/C/ClauJi00.htm',
 'https://www.pro-football-reference.com/players/M/ManzJo00.htm',
 'https://www.pro-football-reference.com/players/S/ShawCo00.htm',
 'https://www.pro-football-reference.com/players/G/GaroJi00.htm',
 'https://www.pro-football-reference.com/players/S/SavaTo00.htm',
 'https://www.pro-football-reference.com/players/T/ThomLo00.htm',
 'https://www.pro-football-reference.com/players/N/NassRy00.htm',
 'https://www.pro-football-reference.com/players/P/PalmJo20.htm',
 'https://www.pro-football-reference.com/players/M/MartSa01.htm',
 'https://www.pro-football-reference.com/players/S/ShorCe00.htm',
 'https://www.pro-football-reference.com/players/A/AngeBr00.htm',
 'https://www.pro-football-reference.com/players/B/BeckOd00.htm',
 'https://www.pro-football-reference.com/players/B/BoldAn00.htm',
 'https://www.pro-football-reference.com/players/C/ColqDu20.htm',
 'https://www.pro-football-reference.com/players/F/FostAr00.htm',
 'https://www.pro-football-reference.com/players/G/GarcPi00.htm',
 'https://www.pro-football-reference.com/players/G/GinnTe00.htm',
 'https://www.pro-football-reference.com/players/G/GoldRo00.htm',
 'https://www.pro-football-reference.com/players/G/GordJo02.htm',
 'https://www.pro-football-reference.com/players/G/GricMa00.htm',
 'https://www.pro-football-reference.com/players/J/JackFr02.htm',
 'https://www.pro-football-reference.com/players/K/KearJe01.htm',
 'https://www.pro-football-reference.com/players/L/LechSh20.htm',
 'https://www.pro-football-reference.com/players/L/LeexAn20.htm',
 'https://www.pro-football-reference.com/players/M/McAfPa44.htm',
 'https://www.pro-football-reference.com/players/Q/QuigRy00.htm',
 'https://www.pro-football-reference.com/players/W/WallMi00.htm',
 'https://www.pro-football-reference.com/players/W/WeddEr99.htm',
 'https://www.pro-football-reference.com/players/W/WingBr00.htm',
 'https://www.pro-football-reference.com/players/W/WinsJa00.htm',
 'https://www.pro-football-reference.com/players/M/MariMa01.htm',
 'https://www.pro-football-reference.com/players/M/McCaA.00.htm',
 'https://www.pro-football-reference.com/players/M/MoorKe02.htm',
 'https://www.pro-football-reference.com/players/J/JoneLa02.htm',
 'https://www.pro-football-reference.com/players/O/OrloDa00.htm',
 'https://www.pro-football-reference.com/players/T/TannAl00.htm',
 'https://www.pro-football-reference.com/players/M/MannSe00.htm',
 'https://www.pro-football-reference.com/players/R/RenfSe00.htm',
 'https://www.pro-football-reference.com/players/D/DaniB.00.htm',
 'https://www.pro-football-reference.com/players/A/AmenDa00.htm',
 'https://www.pro-football-reference.com/players/A/AndrAn00.htm',
 'https://www.pro-football-reference.com/players/C/CadeTr00.htm',
 'https://www.pro-football-reference.com/players/C/CrowJa00.htm',
 'https://www.pro-football-reference.com/players/G/GrimJo01.htm',
 'https://www.pro-football-reference.com/players/H/HogaCh00.htm',
 'https://www.pro-football-reference.com/players/K/KochSa20.htm',
 'https://www.pro-football-reference.com/players/L/LandJa00.htm',
 'https://www.pro-football-reference.com/players/S/SantCa01.htm',
 'https://www.pro-football-reference.com/players/W/WaltBr00.htm']

In [4]:
# We will scrape the 5 players at random in the list of links
for player in random.sample(links, 5):
 
    # pass the url to the position parser
    stats = PlayerParser.PlayerParser().passing(player)
 
    # concat the results with our catch-all dataframe
    all_qb = pd.concat([all_qb, stats], axis=0)

trying  https://www.pro-football-reference.com/players/M/McGlMa01.htm
trying  https://www.pro-football-reference.com/players/O/OrtoKy00.htm
trying  https://www.pro-football-reference.com/players/M/McCoCo00.htm
trying  https://www.pro-football-reference.com/players/G/GaroJi00.htm
trying  https://www.pro-football-reference.com/players/M/McFaDa00.htm


UnboundLocalError: local variable 'rush_df' referenced before assignment

In [13]:
all_qb

Name  Year   Age Throws  Height  Weight  DOB_mo  DOB_day  \
0   Peyton Manning  1998  22.0  Right      77     230       3       24   
1   Peyton Manning  1999  23.0  Right      77     230       3       24   
2   Peyton Manning  2000  24.0  Right      77     230       3       24   
3   Peyton Manning  2001  25.0  Right      77     230       3       24   
4   Peyton Manning  2002  26.0  Right      77     230       3       24   
5   Peyton Manning  2003  27.0  Right      77     230       3       24   
6   Peyton Manning  2004  28.0  Right      77     230       3       24   
7   Peyton Manning  2005  29.0  Right      77     230       3       24   
8   Peyton Manning  2006  30.0  Right      77     230       3       24   
9   Peyton Manning  2007  31.0  Right      77     230       3       24   
10  Peyton Manning  2008  32.0  Right      77     230       3       24   
11  Peyton Manning  2009  33.0  Right      77     230       3       24   
12  Peyton Manning  2010  34.0  Right      77     230       3       24   
13  Peyton Manning  2012  36.0  Right      77     230       3       24   
14  Peyton Manning  2013  37.0  Right      77     230       3       24   
15  Peyton Manning  2014  38.0  Right      77     230       3       24   
16  Peyton Manning  2015  39.0  Right      77     230       3       24   
0        Matt Ryan  2008  23.0  Right      76     217       5       17   
1        Matt Ryan  2009  24.0  Right      76     217       5       17   
2        Matt Ryan  2010  25.0  Right      76     217       5       17   
3        Matt Ryan  2011  26.0  Right      76     217       5       17   
4        Matt Ryan  2012  27.0  Right      76     217       5       17   
5        Matt Ryan  2013  28.0  Right      76     217       5       17   
6        Matt Ryan  2014  29.0  Right      76     217       5       17   
7        Matt Ryan  2015  30.0  Right      76     217       5       17   
8        Matt Ryan  2016  31.0  Right      76     217       5       17   
9        Matt Ryan  2017  32.0  Right      76     217       5       17   
0       Drew Brees  2002  23.0  Right      72     209       1       15   
1       Drew Brees  2003  24.0  Right      72     209       1       15   
2       Drew Brees  2004  25.0  Right      72     209       1       15   
..             ...   ...   ...    ...     ...     ...     ...      ...   
10  Peyton Manning  2008  32.0  Right      77     230       3       24   
11  Peyton Manning  2009  33.0  Right      77     230       3       24   
12  Peyton Manning  2010  34.0  Right      77     230       3       24   
13  Peyton Manning  2012  36.0  Right      77     230       3       24   
14  Peyton Manning  2013  37.0  Right      77     230       3       24   
15  Peyton Manning  2014  38.0  Right      77     230       3       24   
16  Peyton Manning  2015  39.0  Right      77     230       3       24   
0   Donovan McNabb  1999  23.0  Right      74     240      11       25   
1   Donovan McNabb  2000  24.0  Right      74     240      11       25   
2   Donovan McNabb  2001  25.0  Right      74     240      11       25   
3   Donovan McNabb  2002  26.0  Right      74     240      11       25   
4   Donovan McNabb  2003  27.0  Right      74     240      11       25   
5   Donovan McNabb  2004  28.0  Right      74     240      11       25   
6   Donovan McNabb  2005  29.0  Right      74     240      11       25   
7   Donovan McNabb  2006  30.0  Right      74     240      11       25   
8   Donovan McNabb  2007  31.0  Right      74     240      11       25   
9   Donovan McNabb  2008  32.0  Right      74     240      11       25   
10  Donovan McNabb  2009  33.0  Right      74     240      11       25   
11  Donovan McNabb  2010  34.0  Right      74     240      11       25   
12  Donovan McNabb  2011  35.0  Right      74     240      11       25   
0      Jeff Garcia  1999  29.0  Right      73     195       2       24   
1      Jeff Garcia  2000  30.0  Right      73     195       2       24   
2      Jeff Garcia

In [4]:
all_qb.groupby(['Name']).sum()

Year    Age  Height  Weight  DOB_mo  DOB_day  DOB_yr  \
Name                                                                      
Drew Brees        32152  488.0    1152    3344      16      240   31664   
Matt Ryan         20125  275.0     760    2170      50      170   19850   
Matthew Stafford  18117  225.0     675    1980      18       63   17892   
Peyton Manning    34106  514.0    1309    3910      51      408   33592   
Tom Brady         32145  513.0    1216    3600     128       48   31632   

                     Cmp     Att    Cmp%  ...   Tgt  Rec  Rec_Yds   Y/R  \
Name                                      ...                             
Drew Brees        6207.0  9267.0  1066.3  ...   8.0  7.0     73.0  68.0   
Matt Ryan         3630.0  5593.0   646.2  ...   NaN  NaN      NaN   NaN   
Matthew Stafford  3005.0  4850.0   553.0  ...   3.0  2.0     -3.0  -3.0   
Peyton Manning    6125.0  9380.0  1107.4  ...   1.0  1.0     -2.0  -2.0   
Tom Brady         5621.0  8791.0  1023.6  ...   2.0  2.0     59.0  59.0   

                  Rec_TD  Rec_Lng  R/G  Rec_Y/G  YScm  RRTD  
Name                                                         
Drew Brees           1.0     71.0  0.6      5.2   791    19  
Matt Ryan            NaN      NaN  NaN      NaN  1023     5  
Matthew Stafford     0.0     -3.0  0.2     -0.2   946    14  
Peyton Manning       0.0     -2.0  0.1     -0.1   665    18  
Tom Brady            0.0     59.0  0.2      3.8  1027    17  

[5 rows x 46 columns]

<a id='multi_player_multi_season'></a>
###  Multiple player stats for multiple seasons
If you need to get players from multiple seasons, it is easiest to make a pruned list of players to avoid overlap. Then just literate through the list like above.

In [9]:
big_list = []
for year in range(2015,2017):
    big_list = big_list + GetPositionLinks.Position('passing').player_links(year)

In [10]:
# this will contain duplicate urls
len(big_list)

182

In [11]:
# remove the duplicate urls
pruned_list = list(dict.fromkeys(big_list))

In [12]:
len(pruned_list)

129

You would then simply iterate through the "pruned_list" and call:
```python
PlayerParser.PlayerParser().passing(url)
``` 
The result would be a DataFrame of 129 different passing players. 

**Again, this will generate a ton of traffic to the website so use caution.**
<a id='gamelog'></a>
### Gamelog
Individual player gamelog stats can be obtained for a player(s) for a given season(s). Descriptive information about the player such as their alma mater, height, weight, etc is also attached to the results.

In [13]:
from profootballref.LinkBuilder import GetPositionLinks 
from profootballref.Parsers import GamelogParser

# gather player urls for a given season
position = 'passing'
season = 2017

urls = GetPositionLinks.Position(position).player_links(season)

# view the first url as a string
urls[:1][0]

'https://www.pro-football-reference.com/players/B/BradTo00.htm'

In [14]:
# pass the url and the season to the passing method in the GameLog class.
GamelogParser.GameLog().passing(urls[:1][0], season)

Name Pos  Height  Weight  DOB_mo  DOB_day  DOB_yr   College  \
0   Tom Brady  QB      76     225       8        3    1977  Michigan   
1   Tom Brady  QB      76     225       8        3    1977  Michigan   
2   Tom Brady  QB      76     225       8        3    1977  Michigan   
3   Tom Brady  QB      76     225       8        3    1977  Michigan   
4   Tom Brady  QB      76     225       8        3    1977  Michigan   
5   Tom Brady  QB      76     225       8        3    1977  Michigan   
6   Tom Brady  QB      76     225       8        3    1977  Michigan   
7   Tom Brady  QB      76     225       8        3    1977  Michigan   
8   Tom Brady  QB      76     225       8        3    1977  Michigan   
9   Tom Brady  QB      76     225       8        3    1977  Michigan   
10  Tom Brady  QB      76     225       8        3    1977  Michigan   
11  Tom Brady  QB      76     225       8        3    1977  Michigan   
12  Tom Brady  QB      76     225       8        3    1977  Michigan   
13  Tom Brady  QB      76     225       8        3    1977  Michigan   
14  Tom Brady  QB      76     225       8        3    1977  Michigan   
15  Tom Brady  QB      76     225       8        3    1977  Michigan   

          Date    G#        Age   Tm   Home  Opp Result    GS    GS  PF  PA  \
0   2017-09-07   1.0  40.095890  NWE   True  KAN      L  True  True  27  42   
1   2017-09-17   2.0  40.123288  NWE  False  NOR      W  True  True  36  20   
2   2017-09-24   3.0  40.142466  NWE   True  HOU      W  True  True  36  33   
3   2017-10-01   4.0  40.161644  NWE   True  CAR      L  True  True  30  33   
4   2017-10-05   5.0  40.172603  NWE  False  TAM      W  True  True  19  14   
5   2017-10-15   6.0  40.200000  NWE  False  NYJ      W  True  True  24  17   
6   2017-10-22   7.0  40.021918  NWE   True  ATL      W  True  True  23   7   
7   2017-10-29   8.0  40.238356  NWE   True  LAC      W  True  True  21  13   
8   2017-11-12   9.0  40.276712  NWE  False  DEN      W  True  True  41  16   
9   2017-11-19  10.0  40.295890  NWE  False  OAK      W  True  True  33   8   
10  2017-11-26  11.0  40.315068  NWE   True  MIA      W  True  True  35  17   
11  2017-12-03  12.0  40.334247  NWE  False  BUF      W  True  True  23   3   
12  2017-12-11  13.0  40.035616  NWE  False  MIA      L  True  True  20  27   
13  2017-12-17  14.0  40.372603  NWE  False  PIT      W  True  True  27  24   
14  2017-12-24  15.0  40.391781  NWE   True  BUF      W  True  True  37  16   
15  2017-12-31  16.0  40.041096  NWE   True  NYJ      W  True  True  26   6   

    pass_cmp  pass_att   Cmp%  pass_yds  pass_td  Int   Rate  Sk  Sk-Yds  \
0         16        36  44.44       267        0    0   70.0   3      20   
1         30        39  76.92       447        3    0  139.6   2      11   
2         25        35  71.43       378        5    0  146.2   5      41   
3         32        45  71.11       307        2    0  104.6   3      14   
4         30        40  75.00       303        1    1   94.1   3      14   
5         20        38  52.63       257        2    1   80.7   0       0   
6         21        29  72.41       249        2    0  121.2   2       8   
7         32        47  68.09       333        1    0   95.4   3      16   
8         25        34  73.53       266        3    0  125.4   1       6   
9         30        37  81.08       340        3    0  132.0   1       8   
10        18        28  64.29       227        4    1  114.1   1       6   
11        21        30  70.00       258        0    1   82.4   3      14   
12        24        43  55.81       233        1    2   59.5   2      10   
13        22        35  62.86       298        1    1   87.6   2      15   
14        21        28  75.00       224        2    1  106.8   2       6   
15        18        37  48.65       190        2    0   82.0   2      12   

    pass_Y/A   AY/A  rush_att  rush_yds  rush_Y/A  rush_TD  Rec_Tgt  Rec_Rec  \
0       7.42   7.42         2         0       0.0        0        0

The gamelog positions you can choose from are:

**passing()**

**receiving()**

**rushing()**

**defense()**

**kicking()**

Each of these parsers will return a Pandas DataFrame object.

<a id='team_stats'></a>
## Team stats
---
<a id='team_offense'></a>
### Team offense stats
Simply pass a season (year) to the **offense()** method in the **TeamStats()** class

In [15]:
from profootballref.Parsers import TeamStats

year = 2015
df = TeamStats.TeamStats().offense(year)

In [16]:
df.head()

Tm     G     PF     Yds     Ply  Y/P    TO    FL   1stD  \
0     Carolina Panthers  16.0  500.0  5871.0  1060.0  5.5  19.0   9.0  357.0   
1     Arizona Cardinals  16.0  489.0  6533.0  1041.0  6.3  24.0  11.0  373.0   
2  New England Patriots  16.0  465.0  5991.0  1050.0  5.7  14.0   7.0  348.0   
3   Pittsburgh Steelers  16.0  423.0  6327.0  1011.0  6.3  28.0   7.0  331.0   
4      Seattle Seahawks  16.0  423.0  6058.0  1035.0  5.9  16.0   8.0  335.0   

   Pass_Cmp  Pass_Att  Pass_Yds  Pass_TD   Int  NY/A  Pass_1stD  Rush_Att  \
0     300.0     501.0    3589.0     35.0  10.0   6.7      197.0     526.0   
1     353.0     562.0    4616.0     35.0  13.0   7.8      237.0     452.0   
2     404.0     629.0    4587.0     36.0   7.0   6.9      230.0     383.0   
3     391.0     590.0    4603.0     26.0  21.0   7.4      207.0     388.0   
4     333.0     489.0    3790.0     34.0   8.0   7.1      190.0     500.0   

   Rush_Yds  Rush_TD  Y/A  Rush_1stD    Pen  Pen_Yds  1stPy   Sc%   TO%  \
0    2282.0     19.0  4.3      136.0  103.0    887.0   24.0  42.9   9.6   
1    1917.0     16.0  4.2       92.0   94.0    758.0   44.0  42.5  11.8   
2    1404.0     14.0  3.7       87.0   96.0    860.0   31.0  43.2   5.7   
3    1724.0     16.0  4.4       91.0   94.0    868.0   33.0  40.5  13.7   
4    2268.0     10.0  4.5      128.0  117.0   1007.0   17.0  42.0   8.6   

      EXP  
0  125.65  
1  161.96  
2  127.68  
3  116.15  
4  132.31

<a id='team_defense'></a>
### Team defense stats
Simply pass a season (year) to the **defense()** method in the **TeamStats()** class

In [17]:
from profootballref.Parsers import TeamStats

year = 2015
df = TeamStats.TeamStats().defense(year)

In [18]:
df.head()

Tm     G     PF     Yds     Ply  Y/P    TO    FL   1stD  \
0    Seattle Seahawks  16.0  277.0  4668.0   947.0  4.9  23.0   9.0  273.0   
1  Cincinnati Bengals  16.0  279.0  5453.0  1032.0  5.3  28.0   7.0  307.0   
2  Kansas City Chiefs  16.0  287.0  5269.0  1037.0  5.1  29.0   7.0  313.0   
3      Denver Broncos  16.0  296.0  4530.0  1033.0  4.4  27.0  13.0  289.0   
4   Minnesota Vikings  16.0  302.0  5510.0  1015.0  5.4  22.0   9.0  318.0   

   Pass_Cmp  Pass_Att  Pass_Yds  Pass_TD   Int  NY/A  Pass_1stD  Rush_Att  \
0     333.0     548.0    3364.0     14.0  14.0   5.8      175.0     362.0   
1     415.0     646.0    3976.0     18.0  21.0   5.8      202.0     344.0   
2     349.0     607.0    3698.0     25.0  22.0   5.7      193.0     383.0   
3     344.0     573.0    3193.0     19.0  14.0   5.1      162.0     408.0   
4     359.0     561.0    3762.0     24.0  13.0   6.2      189.0     411.0   

   Rush_Yds  Rush_TD  Y/A  Rush_1stD    Pen  Pen_Yds  1stPy   Sc%   TO%  \
0    1304.0     10.0  3.6       71.0   94.0    795.0   27.0  29.3  13.2   
1    1477.0      8.0  4.3       74.0  116.0   1063.0   31.0  28.9  15.0   
2    1571.0      7.0  4.1       86.0  110.0    941.0   34.0  27.3  15.3   
3    1337.0     10.0  3.3       81.0  104.0    773.0   46.0  26.9  11.9   
4    1748.0      7.0  4.3       94.0  109.0    875.0   35.0  33.3  11.9   

      EXP  
0   50.54  
1   24.23  
2   69.97  
3  146.71  
4    3.87